File: Entities/CustomSummitCloud.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;
using Celeste.Mod.Entities;
using ChroniaHelper.Utils;

namespace ChroniaHelper.Entities;

[Tracked(false)]
[CustomEntity("ChroniaHelper/CustomSummitCloud")]

public class CustomSummitCloud : Entity
{
    public Image image;

    public float diff;

    public Vector2 RenderPosition()
    {
        Vector2 camCenter = (base.Scene as Level).Camera.Position + new Vector2(screenX, screenY);
        
        return camCenter + (Position - camCenter) * diff;// parallax

        // Known paramaters : Position and camCenter
        
    }

    private enum Aligns
    {
        PositionTopLeft, PositionTopCenter, PositionTopRight,
        PositionCenterLeft, PositionCenter, PositionCenterRight,
        PositionBottomLeft, PositionBottomCenter, PositionBottomRight,
        AlignTopLeft, AlignTopCenter, AlignTopRight,
        AlignCenterLeft, AlignCenter, AlignCenterRight,
        AlignBottomLeft, AlignBottomCenter, AlignBottomRight
    }

    private Aligns alignment;

    private float floatiness, parallax, screenX, screenY;
    private string path;
    private bool randomizeParallax;

    public CustomSummitCloud(EntityData data, Vector2 offset)
        : base(data.Position + offset)
    {
        base.Tag = Tags.Persistent;
        base.Depth = data.Int("depth", -10550);
        randomizeParallax = data.Bool("randomizeParallax", true);
        parallax = data.Float("parallax");
        screenX = data.Float("screenPosX", 160f);
        screenY = data.Float("screenPosY", 90f);

        diff = randomizeParallax? Calc.Random.Range(parallax, parallax + 0.1f) : parallax;
        floatiness = data.Float("floatiness", 1f);
        alignment = data.Enum<Aligns>("alignment", Aligns.PositionTopLeft);

        List<MTexture> atlasSubtextures = GFX.Game.GetAtlasSubtextures(path = data.Attr("path", "ChroniaHelper/CustomSummitClouds/cloud"));
        image = new Image(Calc.Random.Choose(atlasSubtextures));
        // Alignments
        float tileL = Width, tileH = Height;
        switch (alignment)
        {
            case Aligns.PositionTopCenter:
                image.CenterOrigin();
                Position.X += tileL / 2;
                break;
            case Aligns.PositionTopRight:
                image.CenterOrigin();
                Position.X += tileL;
                break;
            case Aligns.PositionCenterLeft:
                image.CenterOrigin();
                Position.Y += tileH / 2;
                break;
            case Aligns.PositionCenter:
                image.CenterOrigin();
                Position.Y += tileH / 2; Position.X += tileL / 2;
                break;
            case Aligns.PositionCenterRight:
                image.CenterOrigin();
                Position.X += tileL; Position.Y += tileH / 2;
                break;
            case Aligns.PositionBottomLeft:
                image.CenterOrigin();
                Position.Y += tileH;
                break;
            case Aligns.PositionBottomCenter:
                image.CenterOrigin();
                Position.Y += tileH; Position.X += tileL / 2;
                break;
            case Aligns.PositionBottomRight:
                image.CenterOrigin();
                Position.X += tileL; Position.Y += tileH;
                break;
            case Aligns.AlignTopLeft:
                image.JustifyOrigin(0, 0);
                break;
            case Aligns.AlignTopCenter:
                image.JustifyOrigin(0.5f, 0);
                break;
            case Aligns.AlignTopRight:
                image.JustifyOrigin(1f, 0);
                break;
            case Aligns.AlignCenterLeft:
                image.JustifyOrigin(0, 0.5f);
                break;
            case Aligns.AlignCenter:
                image.JustifyOrigin(0.5f, 0.5f);
                break;
            case Aligns.AlignCenterRight:
                image.JustifyOrigin(1f, 0.5f);
                break;
            case Aligns.AlignBottomLeft:
                image.JustifyOrigin(0, 1f);
                break;
            case Aligns.AlignBottomCenter:
                image.JustifyOrigin(0.5f, 1f);
                break;
            case Aligns.AlignBottomRight:
                image.JustifyOrigin(1f, 1f);
                break;
            default:
                image.CenterOrigin();
                break;
        }

        // Scale Flip
        if(data.Bool("randomFlipX", true))
        {
            image.Scale.X = Calc.Random.Choose(-1, 1);
        }
        if(data.Bool("randomFlipY", false))
        {
            image.Scale.Y = Calc.Random.Choose(-1, 1);
        }

        image.Color = Calc.HexToColor(data.Attr("color", "ffffff"));

        Add(image);
        SineWave sineWave = new SineWave(Calc.Random.Range(0.05f, 0.15f) * floatiness, 0f);
        sineWave.Randomize();
        sineWave.OnUpdate = delegate (float f)
        {
            image.Y = f * 8f;
        };
        Add(sineWave);

        // new params
        base.Collider = new Hitbox(8f, 8f);
    }

    public override void Render()
    {
        Vector2 position = Position;
        Position = RenderPosition();
        base.Render();
        Position = position;
    }


}

File: ModFolder/Loenn/entities/CustomSummitCloud.lua
local drawableRectangle = require("structs.drawable_rectangle")
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")
local connectedEntities = require("helpers.connected_entities")
local ChroniaHelper = require("mods").requireFromPlugin("helpers.chroniaHelper_old")
local fo = require("mods").requireFromPlugin("helpers.field_options")
local vivUtils = require("mods").requireFromPlugin("helpers.vivUtilsMig")

local cloud = {}

cloud.name = "ChroniaHelper/CustomSummitCloud"

cloud.depth = function(room,entity) return entity.depth or -10550 end
cloud.placements = {
	name = "cloud",
	data = {
        width = 8,
        height = 8,
        path = "ChroniaHelper/CustomSummitClouds/cloud",
        color = "ffffff",
        floatiness = 1.0,
        depth = -10550,
        parallax = 1.0,
        randomizeParallax = true,
        randomFlipX = true,
        randomFlipY = false,
        alignment = "PositionTopLeft",
        screenPosX = 160,
        screenPosY = 90,
	},
}

cloud.fieldInformation = {
    depth = fo.depths,
    color = {
        fieldType = "color",
    },
    alignment = {
        options = {
            "PositionTopLeft", "PositionTopCenter", "PositionTopRight",
            "PositionCenterLeft", "PositionCenter", "PositionCenterRight",
            "PositionBottomLeft", "PositionBottomCenter", "PositionBottomRight",
            "AlignTopLeft", "AlignTopCenter", "AlignTopRight",
            "AlignCenterLeft", "AlignCenter", "AlignCenterRight",
            "AlignBottomLeft", "AlignBottomCenter", "AlignBottomRight"
        },
        editable = false,
    }
}

local picW, picH, dx, dy

function cloud.sprite(room, entity)
    utils.setSimpleCoordinateSeed(entity.x, entity.y)

    local n = string.format("%02d", math.random(0,2))
    local m = math.random(0,2)
    local sprite = drawableSprite.fromTexture(entity.path .. n, entity)

    dx = 0
    dy = 0
    picW = sprite.meta.width
    picH = sprite.meta.height

    if entity.alignment == "PositionTopCenter" then
        dx = entity.width / 2
    elseif entity.alignment == "PositionTopRight" then
        dx = entity.width
    elseif entity.alignment == "PositionCenterLeft" then
        dy = entity.height / 2
    elseif entity.alignment == "PositionCenter" then
        dx = entity.width / 2
        dy = entity.height / 2
    elseif entity.alignment == "PositionCenterRight" then
        dx = entity.width
        dy = entity.height / 2
    elseif entity.alignment == "PositionBottomLeft" then
        dy = entity.height
    elseif entity.alignment == "PositionBottomCenter" then
        dx = entity.width / 2
        dy = entity.height
    elseif entity.alignment == "PositionBottomRight" then
        dx = entity.width
        dy = entity.height
    elseif entity.alignment == "AlignTopLeft" then
        dx = picW / 2
        dy = picH / 2
    elseif entity.alignment == "AlignTopCenter" then
        dy = picH / 2
    elseif entity.alignment == "AlignTopRight" then
        dx = -picW / 2
        dy = picH / 2
    elseif entity.alignment == "AlignCenterLeft" then
        dx = picW / 2
    elseif entity.alignment == "AlignCenterRight" then
        dx = -picW / 2
    elseif entity.alignment == "AlignBottomLeft" then
        dx = picW / 2
        dy = -picH / 2
    elseif entity.alignment == "AlignBottomCenter" then
        dy = -picH / 2
    elseif entity.alignment == "AlignBottomRight" then
        dx = -picW / 2
        dy = -picH / 2
    else
        dx = 0
        dy = 0
    end

    return sprite:addPosition(dx,dy)
end

--[[
--Render code for Version 2
cloud.selection = function(room, entity)
    return require('utils').rectangle(entity.x - picW / 2 + dx, entity.y - picH / 2 + dy, picW, picH)
end
]]

return cloud